src/chrdev.rs

```diff
@@ -30,24 +49,136 @@ impl Builder {
             )
         };
         if res != 0 {
-            return Err(error::Error::from_kernel_errno(res));
+            return Err(Error::from_kernel_errno(res));
+        }
+
+        let mut cdevs = vec![unsafe { mem::zeroed() }; self.file_ops.len()].into_boxed_slice();
+        for (i, file_op) in self.file_ops.iter().enumerate() {
+            unsafe {
+                bindings::cdev_init(&mut cdevs[i], &file_op.0);
+                cdevs[i].owner = &mut bindings::__this_module;
+                let rc = bindings::cdev_add(&mut cdevs[i], dev + i as bindings::dev_t, 1);
+                if rc != 0 {
+                    // TODO: need to clean up the region + cdevs here.
+                    return Err(Error::from_kernel_errno(rc));
+                }
+            }
         }
+
         return Ok(Registration {
             dev,
             count: self.minors.len(),
+            _cdevs: cdevs,
         });
     }
 }
 
 pub struct Registration {
     dev: bindings::dev_t,
     count: usize,
+    _cdevs: Box<[bindings::cdev]>,
```

Comment:
Why `Box<[...]>` and not `Vec`?

Reply:
Because resizing would invalid the pointers and thus corrupt memory, boxed slice makes clear it can't be resized.

Reply:
Ugh, this feels like too much magic. This is in library code so it's less pressing, but I'm not thrilled with the fact that it's possible to write safe(-ish) code that leads to memory unsafety. Should we be throwing `Pin` at the problem?

Or at least add a comment in the unsafe block that initializes this that says you can't put these into a Vec because they can't move?

Reply:
Does just making this `Pin<Box<...>>` work and commnicate the properties we want?

Reply:
I added a comment.